gensim/models/poincare.py

```diff
@@ -0,0 +1,637 @@
+
+
+class PoincareModel(utils.SaveLoad):
+    """
+    Class for training, using and evaluating Poincare Embeddings described in https://arxiv.org/pdf/1705.08039.pdf
+
+    The model can be stored/loaded via its `save()` and `load()` methods, or stored/loaded in the word2vec
+    format via `wv.save_word2vec_format()` and `KeyedVectors.load_word2vec_format()`.
+
+    Note that training cannot be resumed from a model loaded via `load_word2vec_format`, if you wish to train further,
+    use `save()` and `load()` methods instead.
+
+    """
+    def __init__(
+        self, train_file, size=50, alpha=0.1, negative=10, iter=50,
+        workers=1, epsilon=1e-5, burn_in=10, burn_in_alpha=0.01, encoding='utf8', seed=0):
+        """
+        Initialize and train a Poincare embedding model from a file of transitive closure relations.
+
+        Parameters
+        ----------
+        train_file : str
+            Path to tsv file containing relation pairs.
+        size : int, optional
+            Number of dimensions of the trained model, defaults to 50.
+        alpha : float, optional
+            learning rate for training, defaults to 0.1.
+        negative : int, optional
+            Number of negative samples to use, defaults to 10.
+        iter : int, optional
+            Number of iterations (epochs) over the corpus, defaults to 50.
+        workers : int, optional
+            Number of threads to use for training the model, defaults to 1.
+        epsilon : float, optional
+            Constant used for clipping embeddings below a norm of one, defaults to 1e-5.
+        burn_in : int, optional
+            Number of epochs to use for burn-in initialization (0 means no burn-in), defaults to 0.
+        burn_in_alpha : float, optional
+            learning rate for burn-in initialization, defaults to 0.01, ignored if `burn_in` is 0.
+        encoding : str, optional
+            encoding of training file, defaults to utf8.
+        seed : int, optional
+            seed for random to ensure reproducibility, defaults to 0.
+
+        Returns
+        --------
+        PoincareModel instance
+
+        """
+        self.train_file = train_file
+        self.encoding = encoding
+        self.wv = KeyedVectors()
+        self.size = size
+        self.train_alpha = alpha  # Learning rate for training
+        self.burn_in_alpha = burn_in_alpha  # Learning rate for burn-in
+        self.alpha = alpha  # Current learning rate
+        self.negative = negative
+        self.iter = iter
+        self.workers = workers
+        self.epsilon = epsilon
+        self.burn_in = burn_in
+        self.seed = seed
+        self.random = random.Random(seed)
+        self.np_random = np_random.RandomState(seed)
+        self.init_range = (-0.001, 0.001)
+        self.loss_grad = grad(PoincareModel.loss_fn)
+        self.load_relations()
+        self.init_embeddings()
+
+    def load_relations(self):
```

Comment:
Having to initialize the relations from a file seems unfortunate. If the user happens to have a list of pairs in memory, the API gets clunky. I think we should be able to initialize from an iterable of pairs as well (or instead; getting the iterable of pairs from a file is easy and can be done outside of this class).

Reply:
Good point, extracted it out to a `PoincareData` class. The API accepts an iterable now.
This is also in line with other Gensim models.